crates/turbo-tasks-memory/src/memory_backend.rs

```diff
@@ -283,65 +284,72 @@ impl Backend for MemoryBackend {
         )
     }
 
-    fn track_read_task_output(
-        &self,
-        task: TaskId,
-        reader: TaskId,
-        _turbo_tasks: &dyn TurboTasksBackendApi,
-    ) {
-        if task != reader {
-            self.with_task(task, |t| {
-                t.with_output_mut(|output| {
-                    Task::add_dependency_to_current(TaskDependency::TaskOutput(task));
-                    output.track_read(reader);
-                })
-            })
-        }
-    }
-
     fn try_read_task_cell(
         &self,
-        task: TaskId,
+        task_id: TaskId,
         index: CellId,
         reader: TaskId,
-        _turbo_tasks: &dyn TurboTasksBackendApi,
+        turbo_tasks: &dyn TurboTasksBackendApi,
     ) -> Result<Result<CellContent, EventListener>> {
-        if task == reader {
-            Ok(Ok(self.with_task(task, |task| {
-                task.with_cell_mut(index, |cell| cell.read_content_untracked())
+        if task_id == reader {
+            Ok(Ok(self.with_task(task_id, |task| {
+                task.with_cell(index, |cell| cell.read_own_content_untracked())
             })))
         } else {
-            Task::add_dependency_to_current(TaskDependency::TaskCell(task, index));
-            Ok(Ok(self.with_task(task, |task| {
-                task.with_cell_mut(index, |cell| cell.read_content(reader))
-            })))
+            Task::add_dependency_to_current(TaskDependency::TaskCell(task_id, index));
+            self.with_task(task_id, |task| {
+                match task.with_cell_mut(index, |cell| {
+                    cell.read_content(
+                        reader,
+                        move || format!("{task_id} {index}"),
```

Comment:
I pretty sure that should be cheap. In my mind this will be similar to:

```
struct Closure {
  task_id: TaskId, // 8 bytes
  index: usize // 8 bytes
}

impl FnTrait for Closure {
  pub fn call(&self) -> String {
    format!("{task_id} {index}")
  }
}

// inside Event
description: Box<dyn FnTrait>

Box::new(Closure { task_id, index } as dyn FnTrait)
```

So it basically passes 16 bytes on the heap + 8 bytes for a vtable, but only when it's stored in a `Box<dyn ...>`. When it's passes as `impl Fn() -> String` it should only be 16 bytes on the heap, since the vtable and size is statically known.

But just to be sure this is only stored in a Box at all when the `hanging_detection` feature is enabled. Monomorphation should optimize all that away, so I'm pretty sure it's zero overhead when the feature isn't enabled at all...

Reply:
I also expect this would be DCEd.